### DISCUSSION TITLE
Add support for passing `list[tuple[str, object]]` to `data`

### DISCUSSION THREAD

**Discussion by RobertCraigie** [encode/httpx#2330] [Ideas] [answered]
I have to make a `multipart/form-data` request that includes nested objects. My current solution is to preprocess the field names such that they can be correctly deserialised by the server, e.g. `{'foo': {'bar': ...}` turns into the key: `foo[bar]`.

While this solution works for most cases, it does not work for arrays as the format that the server uses looks like this: `documents[][foo]` which means the keys have to be duplicated for each entry in the array. This is not currently possible to support as httpx expects a dictionary to be passed to `data`. Supporting passing a list of two tuples, like `params` does would make this possible.

I'm happy to put up a PR for this if necessary :)

**Comment by adriangb** [comment, marked as answer]
As far as I know there is no standard way to serialize nested objects in multipart/form-data requests. What you are describing corresponds to what Swagger/OpenAPI calls the "deepObject" serialization method: https://swagger.io/docs/specification/serialization/.

The multipart/form-data spec only states that fields may be singular or repeated. You can send a repeated field like: `httpx.post(..., data={"documents": ["doc1", "doc2"], files=...)`. This _will_ result in "duplicate" data being transmitted. You could also do something like  `httpx.post(..., data={"documents": json.dumps( ["doc1", "doc2"]), files=...)` in which case there's no duplicate data but the server has to know that the field is JSON.

**Reply by lovelydinosaur** [reply]
> As far as I know there is no standard way to serialize nested objects in multipart/form-data requests.

Indeed.

If you *did* want to support a nested style like `foo[bar]` then you'd need to handle that yourself in usercode.

eg...

```python
def nested_data_encoding(data: dict) -> dict:
    """
    Transforms eg. `{'foo': {'bar': ...}` to `{'foo[bar]': ...}`
    """
    ...  # Do your transform here.


httpx.post(url, data=nested_data_encoding(...))
```

It'd be really helpful to submit any progress you make on that here.

**Reply by falkben** [reply]
I wonder if this should be mentioned in the [Requests Compatibility](https://www.python-httpx.org/compatibility/) page, as `requests` will do the tuple unpacking. https://requests.readthedocs.io/en/latest/api/#requests.request 

If so desired, I could try to add a documentation PR for that?

**Reply by lovelydinosaur** [reply]
> as requests will do the tuple unpacking

Could you provide an example of the difference in behaviour?

**Reply by falkben** [reply]
The second example here https://requests.readthedocs.io/en/latest/user/quickstart/#more-complicated-post-requests

**Reply by lovelydinosaur** [reply]
Ah, gotcha. Yes I'd be okay with a note on that.

I'd suggest something like a "Nested Data" note, following the note on `None` values in "[Query Parameters](https://www.python-httpx.org/compatibility/#query-parameters)". They seem like somewhat similar cases to me.

**Reply by falkben** [reply]
I've made a PR to add a note about this: https://github.com/encode/httpx/pull/2461

**Comment by mikenerone** [comment]
I'm just mentioning that this discussion conflates two different things: OP phrasing asked for nesting dictionaries in `data` ("`{'foo': {'bar': ...}`"), which the standards indeed do not cover, but the referenced feature in `requests` isn't that, it's just passing all of the keys and values in the form `[("k1", "v1"), ("k2", "v2"), ...]`, which is a much easier thing to implement. IMO, it should be supported, because whether standards-compliant or not, some servers treat the key order in the form post as significant. I'm dealing with one right now that has duplicate keys _and_ where they appear in the payload _is_ significant. My only solution with `httpx` right now is to manually build the form POST payload myself.

**Reply by lovelydinosaur** [reply]
Key ordering is preserved...

```python
>>> httpx.Request('GET', 'https://www.example.com', data={'a': '123', 'b': '456'}).content
b'a=123&b=456'
>>> httpx.Request('GET', 'https://www.example.com', data={'b': '456', 'a': '123'}).content
b'b=456&a=123'
```

> that has duplicate keys and where they appear in the payload is significant

True, we don't support eg. `"b=456&a=123&b=789"`.

> My only solution with httpx right now is to manually build the form POST payload myself.

I guess that's what I'd do as well...

```
content = urllib.parse.urlencode([("b", "456"), ("a", "123"), ("b", "789")])
headers = {
    "Content-Type": "application/x-www-form-urlencoded",
    "Content-Length": str(len(content))
}
```

**Reply by mikenerone** [reply]
@tomchristie So could we have support for the list-of-2-tuples form for convenience? I'm unclear on whether I should open a separate issue for this request because it's already described by the summary of _this_ issue, though the _description_ describes something completely different.
